### PR TITLE
Classify and polygonize a raster

### PR DESCRIPTION
**Pull request by Gustry**
This PR brings 3 new functions : 
* reclassify
* polygonize_band
* reclassify_polygonize (the mix of these 2)

For the reclassify, I found an algorithm here : https://github.com/chiatt/gdal_reclassify
I have copy/paste the code for now. @timlinux Do you want me to re-write it ? The code seems working for now. I wrote some unit tests.

I would like to continue this PR for the landcover and maybe change every IF using a raster (except earthquake).

### CODE REVIEW

File: safe/gis/reclassify_gdal.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    dst_ds.SetGeoTransform(transform)
+    if nodata in ["True", "true", "t", "T", "yes", "Yes", "Y", "y"]:
+        out_band.SetNoDataValue(default)
+
+    out_band.GetStatistics(0, 1)
+    dst_ds.SetProjection(proj)
+
+
+def _classes_to_string(ranges):
+    output_ranges = []
```
Comment (Gustry) [inline safe/gis/reclassify_gdal.py]: Missing docstring

File: safe/gis/reclassify_gdal.py [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+    :return: The file path to the reclassified raster.
+    :rtype: str
+    """
+    temporary_dir = temp_dir(sub_dir='pre-process')
+    output_raster = unique_filename(
+        suffix='-reclassified.tiff', dir=temporary_dir)
+
+    output_classes, input_classes = _classes_to_string(ranges)
+    default = 0
```
Comment (Gustry) [inline safe/gis/reclassify_gdal.py]: Explain why we should avoid 0 in the ranges

**Comment by timlinux** [comment]
Hi @Gustry 

Cool stuff. Definitely don't rewrite it - its great to re-use existing code where possible. One thing I am curious about is how we do multiple classifications. Bear in mind the use case is to reclassify a raster from continuous data into a single class per range e.g.

screenshot_22

So in that case we would reclassify by this scheme:

Original Value | Class
-------------------|----------
0 | 0
0.01 - 0.2 | 1
0.2 - 2 | 2
2.1 - 5 | 3
5.1 - 10 | 4
10 - 99 | 5

If that use case is satisfied by this PR can you provide a small test to illustrate it and then LGTM to merge

**Comment by timlinux** [comment]
See also #2491

**Comment by Gustry** [comment]
Yes, you can already do this kind of classification.
It's included in the unit test already :  https://github.com/inasafe/inasafe/pull/2757/files#diff-a254ca94a99e82265e74f0f479c44fbaR45
I will add some documentations.

That's why I would like to update IF about tsunami for instance. I had some performance issue last time with tsunami raster.

File: safe/gis/reclassify_gdal.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    return ranges.keys(), output_ranges
+
+
+def reclassify(input_raster, ranges, no_data=0):
+    """Reclassify a raster according to some ranges.
+
+    This function is a wrapper for the code from
+    https://github.com/chiatt/gdal_reclassify
+
+    For instance if you want to classify like this table :
```
Comment (Gustry) [inline safe/gis/reclassify_gdal.py]: @timlinux To classify with multiple thresholds

**Comment by timlinux** [comment]
OK great LGTM once tests pass.

**Pull request merged**